liquidity_pool_router_with_payload/src/lib.rs

```diff
@@ -0,0 +1,242 @@
+        deposit_b: Deposit,
+        sigs: Map<Symbol, Signature>,
+    );
+}
+
+pub struct LiquidityPoolRouter;
+
+impl LiquidityPoolRouterTrait for LiquidityPoolRouter {
+    fn deposit(
+        e: Env,
+        to: Identifier,
+        deposit_a: Deposit,
+        deposit_b: Deposit,
+        sigs: Map<Symbol, Signature>,
+    ) {
+        crate::auth::verify(
+            &e,
+            sigs.get_unchecked(symbol!("deposit")).unwrap(),
+            symbol!("deposit"),
+            (&to, &deposit_a, &deposit_b),
+            e.get_current_call_stack().into(),
+        );
+
+        let salt = pool_salt(&e, &deposit_a.token, &deposit_b.token);
+        if !has_pool(&e, &salt) {
+            let pool_id = pool_contract::create_contract(&e, &salt);
+            put_pool(&e, &salt, &pool_id);
+            pool_contract::Client::new(&e, &pool_id).initialize(&deposit_a.token, &deposit_b.token);
+        }
+        let pool_id = get_pool_id(&e, &salt);
+        let pool_client = pool_contract::Client::new(&e, &pool_id);
+
+        let reserves = pool_client.get_rsrvs();
+        let amounts = get_deposit_amounts(
+            deposit_a.desired,
+            deposit_a.min,
+            deposit_b.desired,
+            deposit_b.min,
+            reserves,
+        );
+
+        let mut sigs_a = Map::new(&e);
+        sigs_a.set(
+            symbol!("sig"),
+            sigs.get_unchecked(symbol!("xfer_a")).unwrap(),
+        );
+        let mut sigs_b = Map::new(&e);
+        sigs_b.set(
+            symbol!("sig"),
+            sigs.get_unchecked(symbol!("xfer_b")).unwrap(),
+        );
+
+        token_contract::Client::new(&e, deposit_a.token).xfer(
+            &to,
+            &deposit_a.nonce,
+            &Identifier::Contract(e.get_current_contract()),
+            &amounts.0,
+            &sigs_a,
+        );
+        token_contract::Client::new(&e, deposit_b.token).xfer(
+            &to,
+            &deposit_b.nonce,
+            &Identifier::Contract(e.get_current_contract()),
+            &amounts.1,
+            &sigs_b,
+        );
+
+        pool_client.deposit(&to);
+    }
+}
+
+pub trait PayloadTrait {
+    fn has_sig(e: Env, function: Symbol) -> bool;
+
+    fn payload(
+        e: Env,
+        function: Symbol,
+        args: Vec<RawVal>,
+        callstack: Vec<(BytesN<32>, Symbol)>,
+    ) -> Map<Symbol, (Identifier, SaltedSignaturePayload)>;
+}
+
+pub struct LiquidityPoolRouterPayload;
+
+#[contractimpl]
+impl PayloadTrait for LiquidityPoolRouterPayload {
+    fn has_sig(_e: Env, function: Symbol) -> bool {
+        const DEPOSIT_RAW: u64 = symbol!("deposit").to_raw().get_payload();
+        match function.to_raw().get_payload() {
+            DEPOSIT_RAW => true,
+            _ => false,
+        }
+    }
+
+    fn payload(
+        e: Env,
+        function: Symbol,
+        args: Vec<RawVal>,
+        callstack: Vec<(BytesN<32>, Symbol)>,
+    ) -> Map<Symbol, (Identifier, SaltedSignaturePayload)> {
```

Comment:
This looks nice, pretty close to what I've imagined, albeit a bit verbose. Maybe we could provide a simple implementation for single-signature contracts, while allowing more complex contracts to be flexible.

Reply:
I'm very open to having multiple versions of this, and I actually think that's one of the major strengths of this approach. For example, we could add a function `fn payload_v(e: Env) -> Symbol` (`payload_v` is short for "payload version") that returns an identifier for what payload function to call. We could even make that `fn payload_v(e: Env, function: Symbol) -> Symbol` if we wanted to allow it to vary by function. Then we could have a couple standards at different complexity levels, and the wallet would just invoke the right one based on the return value of `payload_v`.